pkg/loadbalancing/dnslb/impl.go

```diff
@@ -0,0 +1,166 @@
+package dnslb
+
+import (
+	crand "crypto/rand"
+	"crypto/tls"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	mrand "math/rand"
+	"net"
+	"time"
+
+	"github.com/Cloud-Foundations/Dominator/lib/net/util"
+	"github.com/Cloud-Foundations/golib/pkg/log"
+)
+
+type probeResultType struct {
+	err error
+	ip  net.IP
+}
+
+func listToMap(list []net.IP) map[string]net.IP {
+	ipMap := make(map[string]net.IP, len(list))
+	for _, ip := range list {
+		ipMap[ip.String()] = ip
+	}
+	return ipMap
+}
+
+func newLoadBalancer(config Config, backend RecordReadWriter,
+	logger log.DebugLogger) (*LoadBalancer, error) {
+	if config.FQDN == "" {
+		return nil, errors.New("no FQDN specified")
+	}
+	if config.CheckInterval < time.Second*5 {
+		config.CheckInterval = time.Second * 5
+	}
+	if config.TcpPort < 1 {
+		return nil, errors.New("no TCP port number specified")
+	}
+	crandData := make([]byte, 8)
+	if _, err := crand.Read(crandData); err != nil {
+		return nil, err
+	}
+	seed, _ := binary.Varint(crandData)
+	lb := &LoadBalancer{
+		backend: backend,
+		config:  config,
+		logger:  logger,
+		rand:    mrand.New(mrand.NewSource(seed)),
+	}
+	if myIP, err := util.GetMyIP(); err != nil {
+		return nil, err
+	} else {
+		lb.myIP = myIP
+		lb.myStringIP = myIP.String()
+	}
+	go lb.checkLoop()
+	return lb, nil
+}
+
+func (lb *LoadBalancer) checkIP(ip net.IP) probeResultType {
+	addr := fmt.Sprintf("%s:%d", ip, lb.config.TcpPort)
+	conn, err := net.DialTimeout("tcp", addr, lb.config.CheckInterval>>2)
+	if err != nil {
+		return probeResultType{err: err, ip: ip}
+	}
+	defer conn.Close()
+	if lb.config.DoTLS {
+		tlsConn := tls.Client(conn, &tls.Config{InsecureSkipVerify: true})
```

Comment:
I think there is also a timeout setting for the tls hanshake.

Reply:
I don't see where?

Reply:
Used a deadline on the connection.